core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -661,26 +691,33 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       reader: EventLogFileReader,
       scanTime: Long,
       enableOptimizations: Boolean): Unit = {
+    val rootPath = reader.rootPath
     try {
+      val (shouldReload, lastCompactionIndex) = compact(reader)
```

Comment:
Is there anyway to do compaction as a separate task? Otherwise it seems like this could slow down the creation of the app listing, especially if you're not using disk caching or it's the first time the SHS is creating the listing db.

Reply:
I totally agree it's ideal to do so, but it made me thinking the problem too complicated because compaction "modifies" the event log files which "app listing" and "app rebuild" are reading or are having them as a list of files. 

Actually I had to deal with similar thing in `loadDiskStore` and `createInMemoryStore` (there's retry mechanism), but I'm not sure it doesn't get more complicated if we also take app listing into account.

Reply:
I was thinking that instead of doing this inline here, you could just submit a task to `replayExecutor` after the listing data is updated, so that you give other tasks fetching listing data a chance to run before you try compaction. (You'd call `endProcessing()` at the end of that separate task.)

Any reason why that would not work?

Reply:
I guess what you suggest is separating two tasks and applying lock (processing) for each task, especially let all listing tasks submitted first and all compaction tasks submitted later. 

That would work, but we may also want to consider the difference between cleaning logs and compaction - cleaning logs have its own interval and triggered independently, but compaction is triggered conditionally, only when the return value of shouldReload is true. That means we may want to always do compaction for compaction task, irrelevant of status of `processing` - we wouldn't want to skip, but it's not ideal if we make the task be stuck and wait for `processing`. Maybe has to resubmit at the end of task if the task cannot process due to `processing`?

Reply:
Addressed with my own idea in ec3ffd9d37cc68c36918af8682333b22c7b49d3f - please take a look and suggest other idea if you have better idea.